Change the order of assertion generation to be after the user is asked "is this your computer?"

shane-tomlinson opened the pull request:
It started as a simple exercise.  It ended in a yak shave.
- This fixes a problem where it is possible that a user takes more than 2 minutes to answer, making the already generated assertion invalid.
- A lot of general cleanup of the state machine.
- Push all post email selection flows to a new state, "email_valid_and_ready" which decides whether to ask the user the computer ownership status.
- Assertions are generated after the "is this your computer" is asked.
- doEmailChosen is replaced with doGenerateAssertion - there were multiple calls to generate an assertion, not very DRY.
- Fix a problem in network.js where a user who completes user registration should be marked as authenticated.
- rename email_chosen.js to generate_assertion.js to clarify its purpose.

issue #1460

Commit:
Change the order of assertion generation to be after the user is asked "is this your computer?"

* This fixes a problem where it is possible that a user takes more than 2 minutes to answer, making the already generated assertion invalid.
* A lot of general cleanup of the state machine.
* Push all post email selection flows to a new state, "email_valid_and_ready" which decides whether to ask the user the computer ownership status.
* Assertions are generated after the "is this your computer" is asked.
* doEmailChosen is replaced with doGenerateAssertion - there were multiple calls to generate an assertion, not very DRY.
* Fix a problem in network.js where a user who completes user registration should be marked as authenticated.
* rename email_chosen.js to generate_assertion.js to clarify its purpose.

issue #1460

Commit:
Comment updates

lloyd commented:
omg...

lloyd commented on lib/static_resources.js:
```diff
@@ -97,7 +97,7 @@ var dialog_js = und.flatten([
     '/dialog/controllers/primary_user_provisioned.js',
-    '/dialog/controllers/email_chosen.js',
+    '/dialog/controllers/generate_assertion.js',
     '/dialog/controllers/is_this_your_computer.js',
```
verified removed controller wasn't references anywhere else

lloyd commented on resources/static/dialog/controllers/generate_assertion.js:
```diff
@@ -3,7 +3,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
-BrowserID.Modules.EmailChosen = (function() {
+BrowserID.Modules.GenerateAssertion = (function() {
```
I like the name change.  GenerateAssertion is more granular and clear to me

lloyd commented on resources/static/dialog/resources/state.js:
```diff
@@ -123,6 +123,10 @@ BrowserID.State = (function() {
+      // The user is is authenticated with their IdP. Two possibilities exist
+      // for the email - 1) create a new account or 2) add address to the
+      // existing account. If the user is authenticated with BrowserID, #2
+      // will happen. If not, #1.
```
+1 on addition of comments like this

lloyd commented:
this is a massive change.  the restructuring and renaming makes sense to me, and from a read of the code I'm comfortable with it, provided we put extra testing focus on it, which I'll message to QA.

The bug this fixes already hit someone in IRC in production, so I feel like this is worth the risk.

r+

Commit:
Remove duplicate reset_password through to email validation test.

Update remaining reset_password test to deal with new flow when the user confirms the address.
